Add ARM_MUSCA_A1 target

mikisch81 opened the pull request:
### Description

This PR replaces #8845 as the proting PR of ARM Musca A1.

These targets are defined:
- ARM_MUSCA_A1 as the base target with common code for Secure/Non-secure worlds.
- ARM_MUSCA_A1_NS as the non-secure target.
- ARM_MUSCA_A1_S as the secure target.
  - This target is built using the **bare-metal profile** as described in [ISGDEVPREQ-1003](https://jira.arm.com/browse/ISGDEVPREQ-1003).

- So to build an overall Musca_a1 image containing secure and non-secure with a pre-compiled bootloader just run these commands:
```sh
> mbed compile -t $TOOLCHAIN -m ARM_MUSCA_A1_S --app-config $MBED_OS_ROOT/tools/psa/tfm/mbed_app.json
> mbed compile -t $TOOLCHAIN -m ARM_MUSCA_A1_NS
```

To build special SPM IPC tests do as follows:
```sh
mbed test --compile -t GCC_ARM -m ARM_MUSCA_A1_S --app-config mbed-os/tools/psa/tfm/mbed_app.json -n mbed-os-tests-psa-spm_smoke -DUSE_PSA_TEST_PARTITIONS -DUSE_SMOKE_TESTS_PART1
cd mbed-os/targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/TARGET_MUSCA_A1_NS/device && cp spm_smoke.bin tfm.bin && cd -
mbed test --compile -t GCC_ARM -m ARM_MUSCA_A1_NS -n mbed-os-tests-psa-spm_smoke -DNO_GREENTEA

mbed test --compile -t GCC_ARM -m ARM_MUSCA_A1_S --app-config mbed-os/tools/psa/tfm/mbed_app.json -n mbed-os-tests-psa-spm_client -DUSE_PSA_TEST_PARTITIONS -DUSE_CLIENT_TESTS_PART1
cd mbed-os/targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/TARGET_MUSCA_A1_NS/device && cp spm_client.bin tfm.bin && cd -
mbed test --compile -t GCC_ARM -m ARM_MUSCA_A1_NS -n mbed-os-tests-psa-spm_client -DNO_GREENTEA

mbed test --compile -t GCC_ARM -m ARM_MUSCA_A1_S --app-config mbed-os/tools/psa/tfm/mbed_app.json -n mbed-os-tests-psa-spm_server -DUSE_PSA_TEST_PARTITIONS -DUSE_SERVER_TESTS_PART1 -DUSE_SERVER_TESTS_PART2 --profile debug
cd mbed-os/targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/TARGET_MUSCA_A1_NS/device && cp spm_server.bin tfm.bin && cd -
mbed test --compile -t GCC_ARM -m ARM_MUSCA_A1_NS -n mbed-os-tests-psa-spm_server -DNO_GREENTEA
```

#### Full run of Greentea tests hasn't been run yet, unitesting it by running the mbed-os-tests-psa-its and mbed-os-tests-mbed-crypto-sanity test suites.

#### Currently IAR is not supported.

#### TODO list

[ ] Full run of Greentea tests

### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [] Refactor
    [X] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

@gaborkertesz @theotherjimmy @ashok-rao @dreemkiller @deepikabhavnani @bridadan @alzix

Commit:
Add ARM_MUSCA_A1 as a new target platform

Musca-A1 is a Cortex-M33 based target with security extension enabled.

- ARM_MUSCA_A1 is the non-secure target running mbed-os.
- ARM_MUSCA_A1_S is the secure target running TF-M.
- TF-M sources were imported and patched in previous commits.
- TF-M secure bootloader (McuBoot) for MUSCA_A1 is submitted by a pre-built binary.
- A post-build hook concatenates The secure and non-secure binaries,
  signs it and then concatenates the bootloader with the signed binary.

jeromecoutant commented:
@jamike

ciarmcom commented:
@mikisch81, thank you for your changes.
@dreemkiller @ashok-rao @theotherjimmy @bridadan @alzix @deepikabhavnani@ARMmbed/mbed-os-maintainers @ARMmbed/mbed-os-tools please review.

theotherjimmy commented:
@mikisch81 Bare metal profile is not planned to handle changing the meaning of build profiles, and I don't think that it's a good idea (it's mixing concerns for 2, currently independent ideas: what code to include and what flags to use).

theotherjimmy commented:
@mikisch81 could you expand on why you need the build profiles? what is different between matching profiles?

mikisch81 commented:
> @mikisch81 could you expand on why you need the build profiles? what is different between matching profiles?

@theotherjimmy I encountered 2 issues when trying to build TF-M secure code using mbed-os c like:
1. All the __wrap__xxx which are added to the linker means that a wrapper for main, malloc, etc. are needed. (Is it preferable to add a file which implements dummy wrappers which call the real functions?).
2. For ARMC6 TF-M use *-std=c99* and mbed-os use *-std=gnu99*.

mikisch81 commented:
@bulislaw @donatieng

deepikabhavnani commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/device/COMPONENT_NSPE/TOOLCHAIN_ARMC6/musca_ns.sct:
```diff
@@ -0,0 +1,37 @@
+
+#include "../cmsis_nvic.h"
+
+LR_CODE NS_CODE_START NS_CODE_SIZE {
```
`NS_CODE_START` and `NS_CODE_SIZE` - Where are these defines set? Are those configurable?

mikisch81 replied:
They are defined in [region_defs.h](https://github.com/ARMmbed/mbed-os/blob/b44fb46c39a831ac94433cfa0b66198ca740f245/targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/partition/region_defs.h), _cmsis_nvic.h_ includes this file.
But to be honest I think I will add an explicit include in the ld/sct files as well (explicit is better than implicit 😄)

mikisch81 replied:
Done

deepikabhavnani replied:
@mikisch81 - We have few common defines (RAM / ROM start and size) already available in the tools, which are picked from CMSIS packages and passed as defines. It will be good to use the same instead of adding new ones. 

Please check in build of MUSCA if `MBED_RAM_START` `MBED_ROM_START` are defined, if yes you can use them

mikisch81 replied:
@deepikabhavnani Yes, I know these defines, we use them in PSOC6.
MUSCA build doesn't define them, it takes the values from `region_defs.h` which is from the TF-M port of Musca.
I can add a block like in [PSOC6](https://github.com/ARMmbed/mbed-os/blob/master/targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8C63XX/TARGET_MCU_PSOC6_M4/device/TOOLCHAIN_GCC_ARM/cy8c6xx7_cm4_dual.ld#L31-L45) which will set these defines accordingly.

deepikabhavnani replied:
Will appreciate if defines in `region_defs.h` and linker files are updated and used as `PSOC6`, it will give flexibility to users instead of hard-coded values.

mikisch81 replied:
I think it should be relevant only to NS side, as the secure side only runs TF-M on bare-metal.
Something like this in the NS linker scripts:
```c
#if !defined(MBED_ROM_START)
  #define MBED_ROM_START    NS_CODE_START
#endif

#if !defined(MBED_ROM_SIZE)
  #define MBED_ROM_SIZE     ((TOTAL_CODE_SRAM_SIZE / 2) - BL2_HEADER_SIZE)
#endif

#if !defined(MBED_RAM_START)
  #define MBED_RAM_START    NVIC_RAM_VECTOR_LIMIT
#endif

#if !defined(MBED_RAM_SIZE)
  #define MBED_RAM_SIZE     (NS_DATA_SIZE - NVIC_RAM_VECTOR_SIZE)
#endif
```

deepikabhavnani replied:
TF-M though a bare-metal layer, it has linker files https://github.com/ARMmbed/mbed-os/pull/9221/files#diff-477bb268cd4ef63fbe0f6f161bbfeb7cR25

Change is relevant for secure + non-secure side as well. Idea is to have defines at one place, instead of hard-coded in code, so if someone wants to change the size of secure / non-secure RAM / ROM it all should be at one place. Now it looks like it is `region_defs.h`  

I will propose something like this instead

```
#if !defined(MBED_ROM_START)
  #define MBED_ROM_START    0x10200000
#endif

#if !defined(MBED_ROM1_START)
  #define MBED_ROM1_START    0x00200000
#endif

#define S_ROM_ALIAS_BASE           MBED_ROM_START
#define NS_ROM_ALIAS_BASE        MBED_ROM1_START
#define TOTAL_ROM_SIZE      (MBED_ROM_SIZE + MBED_ROM1_SIZE)
```

MBED_ROM_START / MBED_ROM_SIZE - Will map to IROM1
MBED_ROM1_START / MBED_ROM1_SIZE - Will map to IROM2
and same is for RAM

![image](https://user-images.githubusercontent.com/28542450/50661682-788d0000-0f69-11e9-913b-3c46dad471db.png)

deepikabhavnani commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/device/COMPONENT_SPE/TOOLCHAIN_ARMC6/musca_s.sct:
```diff
@@ -0,0 +1,211 @@
+LR_CODE S_CODE_START {
+
+    /****  This initial section contains common code for TEE */
+    ER_TFM_CODE S_CODE_START S_CODE_SIZE {
```
`S_CODE_START` and `S_CODE_SIZE` - From where are these values parsed?

mikisch81 replied:
See [this answer](https://github.com/ARMmbed/mbed-os/pull/9221#discussion_r244934535)

jeromecoutant replied:
Shouldn't we use value from targets.json inherited from PSA_Target:
         "overrides": {
            "secure-rom-start": "xx",
            "secure-rom-size": "xx",
            "non-secure-rom-start": "xx",
            "non-secure-rom-size": "xx",
            "secure-ram-start": "xx",
            "secure-ram-size": "xx",
            "non-secure-ram-start": "xx",
            "non-secure-ram-size": "xx",
            "shared-ram-start": "xx",
            "shared-ram-size": "xx"
        }

mikisch81 replied:
@jeromecoutant This comment is outdated and not relevant anymore.

deepikabhavnani commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/device/COMPONENT_SPE/TOOLCHAIN_ARMC6/musca_s.sct:
```diff
@@ -0,0 +1,211 @@
+        *(TFM_SP_STORAGE_ATTR_FN)
+    }
+
+    TFM_SP_AUDIT_LOG +0 ALIGN 32 {
```
Since audit log and storage will not be part of Mbed-OS, linker script should be updated as part of patch to remove these features

mikisch81 replied:
You are right, the LD files are generated in TF-M but the SCT files are not, need to manually change them.

mikisch81 replied:
Done

deepikabhavnani commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/device/COMPONENT_SPE/TOOLCHAIN_ARMC6/musca_s.sct:
```diff
@@ -0,0 +1,211 @@
+
+}
+
+LR_VENEER CMSE_VENEER_REGION_START {
```
Where is `CMSE_VENEER_REGION_START ` defined?

mikisch81 replied:
See [this answer](https://github.com/ARMmbed/mbed-os/pull/9221#discussion_r244934535)

deepikabhavnani commented:
Since TFM source is pulled in as well, do we need tfm.bin and cmse_lib.o? Mcuboot.bin might still be needed.

mikisch81 commented:
> 
> 
> Since TFM source is pulled in as well, do we need tfm.bin and cmse_lib.o? Mcuboot.bin might still be needed.

Eventually we will need some default pre-compiled binaries in the tree (for CI), need to agree on the location.
@theotherjimmy

mikisch81 commented:
CC @gaborkertesz

deepikabhavnani commented:
> Eventually we will need some default pre-compiled binaries in the tree (for CI), need to agree on the location.

@ARMmbed/mbed-os-maintainers 
With pre-compiled binaries maintenance might be a issue, any change in secure side of code will not be tested unless binaries are added with it. Not sure if everyone has agreed to it, just wanted to put it here

bulislaw commented:
That's a new platform support, we'll need full test result on all 3 compilers. Why do we need the certificate in the sources? What are the changes to the build profiles and how users are supposed to use the new files? I don't want to force everyone to know about and manually pass path to the files.

deepikabhavnani commented:
> That's a new platform support, we'll need full test result on all 3 compilers

@bulislaw -  IAR is not supported for ARMv8 devices (IAR 8.x dependency)

mikisch81 commented:
@bulislaw 

> Why do we need the certificate in the sources?

Do you mean the private key used by `imgtool.py` to sign the concatenated binary?

> What are the changes to the build profiles and how users are supposed to use the new files? I don't want to force everyone to know about and manually pass path to the files.

I agree, see my [answer](https://github.com/ARMmbed/mbed-os/pull/9221#issuecomment-450945632) to @theotherjimmy, bullet 2 is not relevant anymore as I was able today to build TF-M for ARMC6 with `-std=gnu99`, if there is a more elegant way to deal with bullet 1 I prefer it.

0xc0170 commented:
For .bin and .o files, please add readme with description (useful to have how it was compiled - version, and what it contains. also the info from above, when it should be precompiled, might not be clear to everyone?) and also license file

mikisch81 commented:
Rebased on latest `master` branch

0xc0170 commented:
astyle - master now contains a fix (there was a bug in the recent change) - ignoring folders, that might affect this PR as I noticed in Travis - if you can rebase, should fix it

cyliangtw commented:
@mikisch81, 
I can't pass green-tea test on musca-a1 on hand, how do you proceed, even PSA test ?
In targets.json, ARM_MUSCA_A1_S set TFM_LVL=1, if current code could support Level3, is there any test code or log for Level 2 & Level 3 ?

0xc0170 commented:
Seems this is still not resolved

> All the __wrap__xxx which are added to the linker means that a wrapper for main, malloc, etc. are needed. (Is it preferable to add a file which implements dummy wrappers which call the real functions?).

What is the problem here?

mikisch81 commented:
> 
> 
> Seems this is still not resolved
> 
> > All the __wrap__xxx which are added to the linker means that a wrapper for main, malloc, etc. are needed. (Is it preferable to add a file which implements dummy wrappers which call the real functions?).
> 
> What is the problem here?

@0xc0170 This is resolved by commit eebd1322721aa7eb7b2519e9b233b92fc6575509 which adds a wrapper file which implements all the wrappers which the linker requires, thus no need for a custom build profile for now. @theotherjimmy

theotherjimmy commented:
:tada: No custom build profiles.

mikisch81 commented:
> 
> 
> @mikisch81 Why is this needed - it's not using mbed default one?

@0xc0170 As I am currently not building the entire tree and just a subset using `--source` I need to supply all the wrappers defined in the build profile.
If we will be able to build the whole tree for TF-M we can consider using the default one from the `platform` folder.

From a quick glance at `platform/mbed_alloc_wrappers.cpp` it appears that if `MBED_MEM_TRACING_ENABLED` and `MBED_HEAP_STATS_ENABLED` are not set the behaviour will be the same as the new file I added, right?
Just missing the `main`, `exit` and `at_exit` wrappers.

0xc0170 commented:
Looks like astyle is picking up .sh file? should not cc @cmonr

cmonr commented:
@0xc0170 Thanks for the heads up.
You're right, astyle _is_ picking up a shell script. That's because _it's in the root of Mbed OS_.

Blocking this PR until the script is moved to a more appropriate place (somewhere under `/tools`).

cmonr reviewed (dismissed):
`build_tfm.sh` should be moved outside of the root.

Preferably somewhere in `/tools`. A more proper review will follow.

mikisch81 commented:
@cmonr The shell script will never be merged (it's in a separate commit with a big DO_NOT_MERGE prefix).

It's more as a reference to everyone on how the secure target is currently built, at least until we can start prototyping the new bare-metal profile (@theotherjimmy )

mikisch81 commented:
But it will move to the tools folder at least to let astyle pass..

deepikabhavnani commented:
> It's more as a reference to everyone on how the secure target is currently built, at least until we can start prototyping the new bare-metal profile (@theotherjimmy )

If its reference, it should be a document with reference code of shell script

bridadan commented:
Hi @mikisch81, sorry for my very late reply. It appeared that this PR was in a bit of flux so I held off on reviewing (too long for my liking). Is it reasonable to review this change set now?

mikisch81 commented:
> Hi @mikisch81, sorry for my very late reply. It appeared that this PR was in a bit of flux so I held off on reviewing (too long for my liking). Is it reasonable to review this change set now?

@bridadan I guess you will mainly look at the `tools` changeset.
Well, you can look at it, it is mostly taken from the previous PR, I guess some of the target specific part there can go outside (as it is taken from the TF-M bootloader module).
And there will be more changes (an update of the TF-M sources is due soon).

donatieng commented:
As mentioned by @bulislaw we need `mbed test` results for ARMC6 and GCC.

alzix commented:
@bridadan,

>Hi @mikisch81, sorry for my very late reply. It appeared that this PR was in a bit of flux so I held off on reviewing (too long for my liking). Is it reasonable to review this change set now?

The PR is still in WIP mode.
Following changes expected to land on it:
- TF-M SPM updated to IPC model instead of library model
- reuse PSA crypto and PSA internal storage services from `components/TARGET_PSA/services`
- PSA attestation service added - from more recent TF-M repo
- Musca flash API implemented for working with TDBStore
- the `build_tfm.sh` expected to be replaced by more elegant mbed way of building

mikisch81 commented:
Rebased on top of latest master branch.

Implemented flash HAL driver for Musca_a1_s target and added support for the PITS (@ARMmbed/mbed-os-storage).

mikisch81 commented:
@ARMmbed/mbed-os-maintainers 
**travis-ci/doxy-spellcheck** has a non-related error from `mbed_rtc_time.h`:
```sh
platform/mbed_rtc_time.h
=================================
Errors: 
101:/** retarget
113:/** retarget
103: secondsa
115: secondsa
_________________________________
```

mikisch81 commented:
> 
> 
> As mentioned by @bulislaw we need `mbed test` results for ARMC6 and GCC.

@donatieng You are right, when this PR will be complete we will share greentea tests results for both GCC and ARMC6.

0xc0170 commented:
> travis-ci/doxy-spellcheck has a non-related error from mbed_rtc_time.h:

Rebasing this branch now (to get a fix from master), travis should be green

cmonr commented:
**NOTE:** This PR has now been rebased.

If this was made in error, feel free to force-push your local branch to restore the PR.

bridadan commented on .astyleignore:
```diff
@@ -22,4 +22,5 @@ TESTS/mbed_hal/trng/pithy
 targets
 components/802.15.4_RF
 components/wifi
+components/TARGET_PSA/TARGET_TFM
```
Is this a temporary ignore?

bridadan commented on targets/targets.json:
```diff
@@ -5091,6 +5091,45 @@
+        ],
+        "extra_labels": ["ARM_SSG", "MUSCA_A1", "PSA", "TFM"],
+        "supported_toolchains": ["ARMC6", "GCC_ARM"],
+        "default_toolchain": "ARMC6",
```
I don't believe `ARMC6` is a valid setting for this field. It's pretty confusing, but due to legacy reasons this controls whether a target uses the "microlib" or not when being built in the online compiler (which only supports ARM compiler 5). I would recommend just dropping the line.
```suggestion
```

bridadan commented on tools/build_tfm.sh:
```diff
@@ -0,0 +1,21 @@
+    --source $MBED_OS_ROOT/features/storage/kvstore/include \
+    --source $MBED_OS_ROOT/features/storage/kvstore/conf/tdb_internal \
+    --source $MBED_OS_ROOT/components/TARGET_PSA/services/psa_prot_internal_storage \
+    $*
```
Just a reminder to drop this script when this PR is ready (this was discussed previously I believe)

bridadan commented on tools/resources/__init__.py:
```diff
@@ -148,6 +148,11 @@ def __init__(self, notify, collect_ignores=False):
+        # make sure mbed-os root is added as include directory
+        script_dir = dirname(abspath(__file__))
+        mbed_os_root_dir = normpath(join(script_dir, '..', '..'))
+        self.add_file_ref(FileType.INC_DIR, mbed_os_root_dir, mbed_os_root_dir)
```
Is there a case where not having this breaks the build? I don't believe this should be necessary.

bridadan commented on tools/targets/ARM_MUSCA_A1/ARM_MUSCA_A1_TFM.py:
```diff
@@ -0,0 +1,103 @@
+try:
+ import Crypto
+except ImportError, e:
+ print('Please install "pycryptodome" Python module! ("pip install --user pycryptodome")')
```
Please add `pycryptodome` to the [requirements.txt](https://github.com/ARMmbed/mbed-os/blob/master/requirements.txt) instead of printing this error.

bridadan replied:
Bump

bridadan commented on tools/targets/ARM_MUSCA_A1/assemble.py:
```diff
@@ -0,0 +1,105 @@
+offset_re = re.compile(r"^#define ([0-9A-Z_]+)_IMAGE_OFFSET\s+((0x)?[0-9a-fA-F]+)")
+size_re   = re.compile(r"^#define ([0-9A-Z_]+)_IMAGE_MAX_SIZE\s+((0x)?[0-9a-fA-F]+)")
+
+class Assembly():
```
This whole file seems to have quite a bit of overlap with our concept of "regions" in the build tools. Does that seem to be the case @theotherjimmy? Perhaps we can reduce the amount of target-specific scripts needed?

bridadan commented on tools/targets/ARM_MUSCA_A1/imgtool_lib/image.py:
```diff
@@ -0,0 +1,179 @@
+                self.version.build or 0,
+                0) # Pad2
+        self.payload = bytearray(self.payload)
+        self.payload[:len(header)] = header
```
Once again, seems like quite a bit of overlap with "regions" within the build system. Perhaps we can merge the efforts.

bridadan commented on tools/targets/ARM_MUSCA_A1/imgtool_lib/version.py:
```diff
@@ -0,0 +1,66 @@
+    print(decode_version("1.2"))
+    print(decode_version("1.0"))
+    print(decode_version("0.0.2+75"))
+    print(decode_version("0.0.0+00"))
```
Can these tests be dropped?

bridadan replied:
Bump

bridadan commented on tools/targets/ARM_MUSCA_A1/root-rsa-2048.pem:
```diff
@@ -0,0 +1,27 @@
+y+rnmQKBgCh0gfyBT/OKPkfKv+Bbt8HcoxgEj+TyH+vYdeTbP9ZSJ6y5utMbPg7z
+iLLbJ+9IcSwPCwJSmI+I0Om4xEp4ZblCrzAG7hWvG2NNzxQjmoOOrAANyTvJR/ap
+N+UkQA4WrMSKEYyBlRS/hR9Unz31vMc2k9Re0ukWhWh/QksQGDfJ
+-----END RSA PRIVATE KEY-----
```
Is there anything special about this key? Checking in a private key is generally not a good thing...

bridadan replied:
Could you add some comments about this key?

bridadan commented on tools/targets/ARM_MUSCA_A1/ARM_MUSCA_A1_TFM.py:
```diff
@@ -0,0 +1,103 @@
+    temporary_files = [concatenated_binary_file, signed_binary_file]
+
+    #1. Run assemble.py to concatenate secure TFM and non-secure mbed binaries
+    sys.argv = ['assemble.py',
```
We really do not want to be modifying the `sys.argv` global variable. I'd prefer these options to be passed into the `assemble.main()` function as arguments.

bridadan commented on tools/targets/ARM_MUSCA_A1/ARM_MUSCA_A1_TFM.py:
```diff
@@ -0,0 +1,103 @@
+                'sign', '--layout', flash_layout_file,
+                '-k', join(run_dir, 'root-rsa-2048.pem'),
+                '--align', '1', '-v', '1.0', '-H', '0x400', '--pad', '0x100000',
+                concatenated_binary_file, signed_binary_file]
```
Same thing regarding modifying `sys.argv`

bridadan reviewed (changes requested):
Thanks for your patience! I realize this PR is still in WIP mode, so if any of my comments will be out of date in a new implementation feel free to say so.

0xc0170 commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/us_ticker.c:
```diff
@@ -0,0 +1,141 @@
+    us_ticker_irq_handler();
+}
+
+#else // COMPONENT_NSPE
```
what is this `COMPONENT_NSPE` else here ? I would rather disable entire us ticker support rather than providing empty implementation?

I've noticed it also in pinmap and possibly others - what is the idea with this NSPE?

alzix replied:
@0xc0170,
The idea is that we need to build TDBStore and storage stack as part of TF-M. TF-M can be considered as a bare metal application from mbed-os perspective.
By trying to build storage stack we pulling in big addition sources. Most of them are not used but build fails on missing implementations, missing headers etc.
COMPONENT_NSPE is used to distinguish between secure and regular mbed-os builds.
I agree it looks like ad-hoc solution here.

0xc0170 commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/device/COMPONENT_NSPE/cmsis_nvic_virtual.h:
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2018 Arm Limited
```
in this folder (where is this header file), there are som bin files . They need a license file and readme describing what they are and toolchain versions they were compiled with (find other .hex/.bin in the tree for examples)

mikisch81 commented:
@deepikabhavnani @bridadan @theotherjimmy  last commit 673445e5fc9a50319c905fe3bc4d051b0fbd6d20 adds secure/non-secure RAM/ROM configuration attributes to the Musca targets, we still need to decide if this is the proper way.

bridadan commented:
> last commit 673445e adds secure/non-secure RAM/ROM configuration attributes to the Musca targets, we still need to decide if this is the proper way.

Thanks very much for the heads up @mikisch81, we're keeping the conversation going on this side, hope to have some answers soon.

mikisch81 commented:
Latest commits enable its and psa-crypto to pass.

Currently GCC on Release profile fails, it is under investigation.

mikisch81 commented:
Last 5 commits removes TF-M platform partition and moves its' reset service to our platform partition.
Also adds the TF-M autogen tool and re-generate partitions headers.

jeromecoutant commented on targets/targets.json:
```diff
@@ -7869,7 +7930,7 @@
         "inherits": ["SPE_Target", "FUTURE_SEQUANA_M0"],
         "components_add": ["SPM_MAILBOX", "FLASHIAP"],
-        "extra_labels_add": ["PSA"],
+        "extra_labels_add": ["PSA", "MBED_SPM"],
```
Is it related to this PR ?

mikisch81 replied:
no, the rebase will remove this.

jeromecoutant commented on targets/targets.json:
```diff
@@ -5092,6 +5092,67 @@
+            "TFM_PSA_API",
+            "MBEDTLS_PSA_CRYPTO_C"
+        ],
+        "extra_labels_add": ["MUSCA_A1_NS", "PSA", "TFM"],
```
PSA and TFM should be component ?

mikisch81 replied:
no, they are targets, so build system will include `TARGET_PSA` and `TARGET_TFM` folders.

jeromecoutant replied:
For me, "feature" from "component" directory should be enabled with "component_add", not with "extra_labels-add"
I thought it was mbed-os philosophy ?

0xc0170 replied:
I also question: `components/TARGET_PSA`. Why this is not `COMPONENT_PSA` ? I see components are inside `TARGET_MBED_SPM`.

mikisch81 replied:
@0xc0170 @jeromecoutant We have no objection to changing it to **COMPONENT_**.
But this needs to be in a separated PR, as **TARGET_PSA** is with us since 5.11 release.

deepikabhavnani replied:
@0xc0170 - I have been requesting the reverse, to move Target_PSA to target (as it is target specific code). It was not possible in 5.11 due to build system limitations. But seems possible after 5.12 release of updated tools

@mikisch81 - Correct me here, if still not feasible or has some other limitation.

mikisch81 replied:
@deepikabhavnani what exactly did you request?
Moving the entire **TARGET_PSA** folder into `mbed-os/targets/`?

bridadan replied:
@jeromecoutant @0xc0170 @deepikabhavnani as @mikisch81 said the `components/TARGET_PSA` was brought in for the 5.11, so it is already mainlined. If we want to move or rename this folder (based on the conversation above, that sounds like a good plan anyway), can we move that discussion and work to a separate PR?

0xc0170 replied:
@mikisch81 Can you capture this in the new issue please?

bulislaw replied:
👍

urutva replied:
@0xc0170 Has this already been captured as an issue? If so can you please link it? If not then I can create an issue.

jeromecoutant reviewed (commented):
1 general question about all scripts in tools/targets/ARM_MUSCA_A1
Are they really specific to MUSCA ? Then for each additional supported target, we will need to update these scripts ?

@LMESTM

mikisch81 commented:
> 
> 
> 1 general question about all scripts in tools/targets/ARM_MUSCA_A1
> Are they really specific to MUSCA ? Then for each additional supported target, we will need to update these scripts ?
> 
> @LMESTM

I suggest to wait for the upcoming rebase on master, it will clean this PR alot..

jeromecoutant commented on targets/targets.json:
```diff
@@ -5092,6 +5092,67 @@
+            "BYPASS_NVSTORE_CHECK",
+            "TFM_LVL=1",
+            "DAUTH_CHIP_DEFAULT",
+            "TFM_PSA_API",
```
Maybe some macro can be part of some TFM mbed lib json file ?
Goal is to avoid to repeat the same macro for all targets supporting TFM.

mikisch81 replied:
Yes, you are right, I guess we will find more stuff which can be moved from Musca to a more general place.

mmahadevan108 commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/TARGET_MUSCA_A1_S/device/TOOLCHAIN_ARMC6/musca_s.sct:
```diff
@@ -0,0 +1,151 @@
+        .ANY (+RO)
+    }
+
+#if TFM_LVL == 1
```
Do we need this check given the non-secure configuration is using a different linker script?

mikisch81 replied:
The TFM level doesn't care about the non-secure side, only the partitioning level in the secure side.

mikisch81 commented:
@jeromecoutant 
I am going to rebase this PR on master, it will reduce the number of commits (as they were from merged-PRs).

mikisch81 commented:
PR was rebased on top of latest master branch, all the dependent PRs were merged.

NirSonnenschein commented:
Is this a new target or a refactor? if it is both, maybe if it is both it should be split into 2 PRs?

jeromecoutant commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/partition/flash_layout.h:
```diff
@@ -0,0 +1,151 @@
+ * above.
+ */
+#define FLASH_PARTITION_SIZE            (0x80000) /* 512KB */
+
```
Shouldn't we use value from targets.json inherited from PSA_Target:
"overrides": {
"secure-rom-start": "xx",
"secure-rom-size": "xx",
"non-secure-rom-start": "xx",
"non-secure-rom-size": "xx",
"secure-ram-start": "xx",
"secure-ram-size": "xx",
"non-secure-ram-start": "xx",
"non-secure-ram-size": "xx",
"shared-ram-start": "xx",
"shared-ram-size": "xx"
}

mikisch81 replied:
Yes, we are going to align the partition files with the configuration.

mikisch81 commented:
> 
> 
> Is this a new target or a refactor? if it is both, maybe if it is both it should be split into 2 PRs?

@NirSonnenschein It is a new target.
And it can't be split into 2 PRs, as the Musca_a1 target is the refernce platform for TF-M.
Please note that this PR is used already by SiPs as a reference for their ongoing port of new PSA targets to mbed-os.

mikisch81 commented:
I am currently removing from this PR all the non-Musca related stuff so this PR will only be a **TARGET Update** PR.

This PR will then depend on the other PR.

Will update when finished.

mikisch81 commented:
@jeromecoutant @mmahadevan108 
Please note this PR has been updated and depends on #9653, relevant Musca_a1 porting code is only in commit c3f6cb0ddad031f2b4c1be391fd8d34ebfe12186.

mikisch81 commented:
Rebased on top of #9653

mikisch81 commented:
cc @ARMmbed/team-nuvoton

mikisch81 commented:
`mbed-ls` has `ARM_MUSCA_A1` as a target, as we changed the name of the non-secure target to `ARM_MUSCA_A1_NS` we needed to mock the target:

`mbedls --mock 5006:ARM_MUSCA_A1_NS`

@ARMmbed/mbed-os-tools What do you suggest to do?

bridadan commented:
@mikisch81 there's nothing wrong with continuing to mock the target. However, if you'd like to remove the need to do that you can change the name in mbed-ls to the one present in Mbed OS.

deepikabhavnani commented:
@bridadan - Query : Can we have both names in mbed-ls? Not sure why I remember one target identifies itself with two names, without explicitly mocking

bridadan commented:
> Query : Can we have both names in mbed-ls? Not sure why I remember one target identifies itself with two names, without explicitly mocking

No, `mbed-ls` will give you one platform name per connected device. I believe there was one board that would physically switch the SWD lines to a different MCU. The debugger would pick that up and change the target id of the device accordingly, maybe that's what you were thinking of?

deepikabhavnani commented:
> maybe that's what you were thinking of

No. getting with different command - `mbed detect`.
```

>mbedls
| platform_name | platform_name_unique | mount_point | serial_port | target_id                                        | daplink_version |
|---------------|----------------------|-------------|-------------|--------------------------------------------------|-----------------|
| K64F          | K64F[0]              | D:          | COM36       | 0240000032044e45000a700a997b00356781000097969900 | 0251            |

```
```
>mbed detect
[mbed] Auto-installing missing Python modules...

[mbed] Detected K64F, port COM36, mounted D:, interface version 0251:
[mbed] Supported toolchains for K64F
| Target        | mbed OS 2 | mbed OS 5 | uARM |    IAR    |    ARM    |  GCC_ARM  |
|---------------|-----------|-----------|------|-----------|-----------|-----------|
| K64F          | Supported | Supported |  -   | Supported | Supported | Supported |
| RAPIDIOT_K64F | Supported | Supported |  -   | Supported | Supported | Supported |
Supported targets: 2

```

bridadan reviewed (commented):
I'm concerned about the overlap between the scripts added in `tools/targets/ARM_MUSCA_A1` with the region system we already have for the managed bootloader. It _seems_ like the general idea of the new scripts in `tools/targets/ARM_MUSCA_A` is to merge binaries together and then do some post processing on them (namely signing), is that correct?

bridadan commented:
@deepikabhavnani OH haha well that looks like a bug with the `detect_targets.py` script: https://github.com/ARMmbed/mbed-os/blob/master/tools/detect_targets.py. That'd be good to raise as a separate issue.

mikisch81 commented:
> 
> 
> I'm concerned about the overlap between the scripts added in `tools/targets/ARM_MUSCA_A1` with the region system we already have for the managed bootloader. It _seems_ like the general idea of the new scripts in `tools/targets/ARM_MUSCA_A` is to merge binaries together and then do some post processing on them (namely signing), is that correct?

@bridadan Yes, I am going to push a commit with a refactor of the `tools` stuff, but keep the same logic:
The post-build hook for `TARGET_MUSCA_A1_NS` concatenates the secure and non-secure binaries, then signs the concatenated image using `musca_a1-root-rsa-2048.pem`as the private key, and then prepends the pre-compiled bootloader to the signed concatenated image, like this:
```
1. Concatenate secure TFM and non-secure mbed binaries:
-------------------------------
| Secure bin | Non-secure bin |
-------------------------------

2. Run imgtool to sign the concatenated binary:
--------------------   -------          --------------
| Concatenated bin | + | KEY | ==SIGN=> | Signed bin |
--------------------   -------          --------------

3. Concatenate mcuboot and signed binary and overwrite mbed built binary file:
--------------------
| BL2 | Signed bin |
--------------------
```

The thing is, that the concatenation is not one after the other, but depends on defines from `flash_layout.h`, so the split between the secure and non secure images in the concatenated image will be at `NON_SECURE_IMAGE_OFFSET` (which is 0x80000 on Musca_a1), and the split between the bootloader and the signed image will be at `FLASH_AREA_BL2_SIZE` (which is 0x20000 on Musca_a1).

mikisch81 commented:
@ARMmbed/mbed-os-tools Did a re-factoring of the `tools` folder of this PR, some of the raised issues were addressed in this commit, could you look at it again?

bridadan commented on tools/targets/musca_a1-root-rsa-2048.pem:
```diff
@@ -0,0 +1,27 @@
+y+rnmQKBgCh0gfyBT/OKPkfKv+Bbt8HcoxgEj+TyH+vYdeTbP9ZSJ6y5utMbPg7z
+iLLbJ+9IcSwPCwJSmI+I0Om4xEp4ZblCrzAG7hWvG2NNzxQjmoOOrAANyTvJR/ap
+N+UkQA4WrMSKEYyBlRS/hR9Unz31vMc2k9Re0ukWhWh/QksQGDfJ
+-----END RSA PRIVATE KEY-----
```
I understand that the images need to be signed by _a key_, but could you please explain if there is any significance of _this key_? **It's also a private key, so I'm concerned that it will be committed to the code base.**

mikisch81 replied:
The pre-compiled `mcuboot.bin` which is prepended to the concatenated secure & non-secure iamge, was build with the matching public key of this private key (I guess, it was done by TF-M engineers).
And at boot time this bootloader verifies the signature of the concatenated image using it's public key.
So whenever any target is re-built whether secure or non-secure, the signing process at post-build must happen using this private key.

What other alternative do you suggest?

bridadan replied:
I see. It's just typically in cryptography you don't publish the **private** key. Sounds like this is a limitation of the fact that we're using a pre built mcuboot image? If we were building mcuboot from source each time, a unique key could be generated from our tools before the project was built.

My concern here is that **every binary** built from Mbed OS will be sharing the same private key in this case. Maybe @dreemkiller knows more about this?

bridadan replied:
Maybe the secrecy of this key doesn't matter, but it'd be good to get a confirmation of this! I'm not sure what the expectation of this is.

alzix replied:
This is a test private key. It is required for OOB experience.
We must make sure this private key and its public key counter part can be injected at build time.
I would expect that for production build, OEM will use it's real secret key on his own build machine.
The problematic part for now is that we do have prebuilt binary version of bootloader with a hardcoded public key mathcing this .pem file. Thus it is not easily replaceable.

But it is a good starting point. In next drops we will have TF-M bootloader source level integration. This is the point where we should define a way for our customers to inject custom keypair.

bridadan replied:
Thanks @alzix. Can we add a small doc file next to this key documenting this? Or is there a document saying as much elsewhere?

Perhaps this isn't a big deal if hardware availability for this target is limited? Is that the case?

bridadan commented on tools/tfm/bin_utils/imgtool_lib/keys.py:
```diff
@@ -0,0 +1,104 @@
+from __future__ import print_function
+from Crypto.Hash import SHA256
+from Crypto.PublicKey import RSA
+from Crypto.Signature import PKCS1_v1_5, PKCS1_PSS
```
These functions are coming from the [pycrypto](https://pypi.org/project/pycrypto/) package, correct?

If so, pycrypto is not currently a [dependency](https://github.com/ARMmbed/mbed-os/blob/master/requirements.txt). Could you please either:
- Add it to the requirements.txt file and "pin" the correct major version Ex. `pycrypto>=2.6.1,<3.0` (or whatever is the correct lower bound of the version)
- OR use a module that already comes with Python, if available.

mikisch81 replied:
Actually from `pycryptodome`, I will add it to `requirements.txt`, I don't really know what should be the version range.
Also, all the files under `tools/tfm/bin_utils` were copied as-is from the TF-M repository, do you think they should be added to the importing process in #9653?

bridadan replied:
If you expect them to be updated with new releases from the TF-M repository, then yeah probably can't hurt.

mikisch81 replied:
@bridadan Actually [pycryptodome](https://pypi.org/project/pycryptodome/) should be used as there is an [issue ](https://github.com/dlitz/pycrypto/issues/238) on Windows with python 3.6 and pycrypto, I will add pycryptodome to `requirements.txt`.

bridadan reviewed (commented):
The refactoring definitely looks better!

0xc0170 commented on components/TARGET_PSA/TARGET_TFM/COMPONENT_NSPE/interface/src/tfm_psa_ns_api.c:
```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2018, Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
```
this new component will need a license files: https://os.mbed.com/docs/mbed-os/v5.11/contributing/license.html (follow using different license section please)

0xc0170 replied:
By the time this is ready, there will be list of non apache components in Readme.md, this will need to be added tehre as well.

0xc0170 replied:
My understanding is that this can't be release under Apache 2.0?

mikisch81 replied:
CC @amiraloosh 

There is a ticket opened: anaqua record ID 81133658 which should handle this issue.

Also these are files which are imported in PR #9653, so I guess that PR should also hold for this approval, right?

bridadan reviewed (dismissed):
The only concerns left for me are:

- Missing addition of new python dependencies in [requirements.txt](https://github.com/ARMmbed/mbed-os/blob/master/requirements.txt)
- The publication of the private key
    - Mostly looking for someone to confirm that disclosing this key is ok. If its, we should document why it is ok. Ex. "This is for development or demonstration purposes."

Commit:
Update storage mbed_lib.json for MUSCA_A1_S

0xc0170 commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/PeripheralNames.h:
```diff
@@ -0,0 +1,48 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2018 Arm Limited
```
new files should have 2019

mikisch81 replied:
All files should be 20XX-2019? what is the starting year?

cmonr replied:
New files: 2019
Older files: X-Y   Y is updated to 2019

mikisch81 replied:
Done

0xc0170 commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/TARGET_MUSCA_A1_NS/device/cmsis_nvic.h:
```diff
@@ -0,0 +1,33 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 Arm Limited
```
Can you add SPDX identifiers to new files please

mikisch81 replied:
So all the target files should have SPDX identifiers? (100+ files)

cmonr replied:
Yes

mikisch81 replied:
Done

0xc0170 commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/us_ticker.c:
```diff
@@ -0,0 +1,141 @@
+
+#else // TARGET_MUSCA_A1_NS
+
+void us_ticker_init(void)
```
I believe I already asked but again I do not see anything in this file indicating why the implementation is empty. A comment would be worth.

Still, why `TARGET_MUSCA_A1_NS` is providing empty non functional definition?

0xc0170 replied:
gpio is different, there are comments and implementation empty. Should be the same for us ticker

alzix replied:
Empty implementation is here to satisfy linker, but there is an assumption that this code won't be executed. 
Let's make sure this is the case by adding a call to panic/abort and comment.

jeromecoutant replied:
We think that you should not create any empty function.
Code should be with #if DEVICE_USTICKER,
File will not be compiled thx to targets.json setting ?

alzix replied:
IIRC the are references to this function which are not guarded by any #ifdef guards. @mikisch81, can you double check. 
We agreed we will issue tickets for every such a case. Please make sure we have it covered and add a comment to a ticket to remove this empty implementation

mikisch81 replied:
OK, so when moving `us_ticker.c` into the NS target I get the following linker errors:
```sh
Error: L6218E: Undefined symbol us_ticker_clear_interrupt (referred from BUILD/ARM_MUSCA_A1_S/ARMC6/mbed-os/hal/mbed_us_ticker_api.o).
Error: L6218E: Undefined symbol us_ticker_disable_interrupt (referred from BUILD/ARM_MUSCA_A1_S/ARMC6/mbed-os/hal/mbed_us_ticker_api.o).
Error: L6218E: Undefined symbol us_ticker_fire_interrupt (referred from BUILD/ARM_MUSCA_A1_S/ARMC6/mbed-os/hal/mbed_us_ticker_api.o).
Error: L6218E: Undefined symbol us_ticker_free (referred from BUILD/ARM_MUSCA_A1_S/ARMC6/mbed-os/hal/mbed_us_ticker_api.o).
Error: L6218E: Undefined symbol us_ticker_init (referred from BUILD/ARM_MUSCA_A1_S/ARMC6/mbed-os/hal/mbed_us_ticker_api.o).
Error: L6218E: Undefined symbol us_ticker_read (referred from BUILD/ARM_MUSCA_A1_S/ARMC6/mbed-os/hal/mbed_us_ticker_api.o).
Error: L6218E: Undefined symbol us_ticker_set_interrupt (referred from BUILD/ARM_MUSCA_A1_S/ARMC6/mbed-os/hal/mbed_us_ticker_api.o).
Finished: 0 information, 2 warning and 7 error messages.
```

When looking into `mbed_us_ticker_api.c` I see that there is a static struct of function pointers - [us_interface](https://github.com/ARMmbed/mbed-os/blob/master/hal/mbed_us_ticker_api.c#L23:L32) which the linker will not throw, therefor even that the secure target does not use us_ticker at all, it needs to resolve those symbols - hence the empty implementation.
I agree this is not the right behaviour, but eventhough the target is using the bare-metal prfofile to build, `hal` folder is still built entirely.
@alzix @donatieng @theotherjimmy @bridadan @jeromecoutant

mikisch81 replied:
It appears the problem is that [mbed_die()](https://github.com/ARMmbed/mbed-os/blob/master/platform/mbed_board.c#L26-L49) implementation calls [wait_us()](https://github.com/ARMmbed/mbed-os/blob/master/platform/mbed_wait_api_no_rtos.c#L35-L40) which calls `get_us_ticker_data()` which is defined in `mbed_us_ticker_api.c`.
So just guarding `mbed_us_ticker_api.c` with `#if DEVICE_USTICKER` is not enough.

mikisch81 replied:
Issue #9853 was opened regarding the USTICKER issue.
@ARMmbed/mbed-os-maintainers Can you prioritize this issue, as it is a blocker for porting new secure targets.

0xc0170 replied:
I am reviewing now. us ticker was always taken as "a must" but anyway there was "USTICKER" in device_has.

cc @kjbracey-arm you might be interested in this issue

mikisch81 replied:
@0xc0170 `USTICKER` is only defined for `ARM_MUSCA_A1_NS` target, `ARM_MUSCA_A1_S` does not have it in its' `device_has`. The issue is relevant for secure targets which use TF-M and not Mbed-OS.

0xc0170 replied:
Wasn't this fixed and this empty implementation can be removed?

mikisch81 replied:
Going to reopen #9853, as I still have dependencies with USTICKER.

bridadan reviewed (commented):
Ok thanks very much for adding the requirement.

I've changed my "Request changes" to just "general comment" now.

**My only concern now is the private key!** All we need is some documentation or statement about its intended use!

Commit:
Update target license headers

- Add SPDX identifiers and update dates

Commit:
Add documentation on Musca-A1 signing key

Commit:
Updates after rebase

- postbuild: Use find_secure_image and update prebuilt binaries
- Add CMSIS driver headers
- Align with 2-region memory model

0xc0170 commented on targets/targets.json:
```diff
@@ -5270,6 +5270,61 @@
+    "ARM_MUSCA_A1": {
+        "public": false,
+        "inherits": ["Target"],
+        "supported_toolchains": ["ARMC6", "GCC_ARM"],
```
still no IAR support , just checking the current status as I noticed it's not here .

mikisch81 replied:
Yes, now that IAR 8 is in master we can do it.

0xc0170 commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/TARGET_MUSCA_A1_S/device/drivers/mpu_armv8m_drv.h:
```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2017-2018, Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
```
Dont forget to document  this BSD-3-Clause (https://os.mbed.com/docs/mbed-os/v5.11/contributing/license.html - Using different license).

bridadan commented on tools/targets/ARM_MUSCA_A1.py:
```diff
@@ -0,0 +1,85 @@
+MUSCA_A1_BASE = path_join(MBED_OS_ROOT, 'targets', 'TARGET_ARM_SSG', 'TARGET_MUSCA_A1')
+
+def musca_tfm_bin(t_self, non_secure_bin, secure_bin):
+    tempdir = tempfile.mkdtemp()
```
It would be more appropriate (and more in line with the rest of our targets) to use the BUILD directory instead of a temporary directory for merging of binaries (this is how it is done for managed bootloader).

0xc0170 replied:
@orenc17 has this been answered?

orenc17 replied:
no

bridadan commented on tools/targets/ARM_MUSCA_A1.py:
```diff
@@ -0,0 +1,85 @@
+        with open(non_secure_bin, "w+b") as out_fh:
+            out_fh.write(mcuboot_fh.read())
+            out_fh.seek(mcuboot_image_size)
+            out_fh.write(signed_fh.read())
```
Just as a note to ourselves in the future, we'd like to move this target towards using the "managed bootloader mode" infrastructure to merge these binaries together wherever possible instead of using this custom build hook.

mikisch81 replied:
I partly agree here..  😄 
In the sense that it should not use the "managed bootloader mode" but a "secure bootloader" which is currently a work in progress here.

mikisch81 replied:
Opened internal Jira ticket: [https://jira.arm.com/browse/IOTSEC-1160](https://jira.arm.com/browse/IOTSEC-1160)

bridadan replied:
Ah I'm specifically talking about removing the need for the post build hook merge and using a mainline feature within the tools. But thanks for filing the issue!

LiyouZhou replied:
the tools don't support signing afaik

Commit:
Move us_ticker.c to NS side

Commit:
Build PS tests only if FLASHIAP exists

Commit:
Increase secure heap size to 0x4000

Commit:
Fix secure linker script

Commit:
Increase Secure RAM by 4K

- 68KB Secure
- 60KB Non-secure

Commit:
Add license files

Commit:
Add serial_tx_pinmap() and serial_rx_pinmap() to serial_api.c

Commit:
MUSCA_A1_NS: Add IAR support

Commit:
Fix PinMap_UART_XX

Commit:
Update secure binaries for ARM_MUSCA_A1_S

bridadan commented on targets/targets.json:
```diff
@@ -5345,6 +5345,62 @@
+            "secure-rom-size": "0x7f800",
+            "secure-ram-start": "0x30000000",
+            "secure-ram-size": "0x11000"
+        }
```
No release version for this one? Pretty sure that means this won't be compiled in CI.

mikisch81 replied:
The release version is in the parent target, is it enough?

mikisch81 replied:
LPC55S69 also has this in the parent target so its ok.

bridadan replied:
DOH yup you're right, thanks!

bridadan reviewed (approved):
I have already approved this. I have some old comments that weren't addressed on some of the Python code, though I don't think they block anything.

bridadan reviewed (approved):
Here's another approval @cmonr  😆

gaborkertesz committed:
Fix last issues

Imports working McuBoot for reset.
Updates microsec ticker driver.
Default baudrate is set to 115200 to see TF-M boot messages.
Stack top is set to scatter file dependent and not hard-coded.

0xc0170 commented on tools/toolchains/arm.py:
```diff
@@ -547,14 +547,15 @@ def __init__(self, target, *args, **kwargs):
 
         # FPU handling
-        if core == "Cortex-M4" or core == "Cortex-M7" or "core" == "Cortex-M33":
+        if core in ["Cortex-M4", "Cortex-M7", "Cortex-M33", "Cortex-M33E"]:
```
isn't this on current master (does this PR needs rebase) ?

0xc0170 reviewed (commented):
Github does not want to load me all comments, was IAR tested and results shared?

Commit:
Use build directory in postbuild

Use build directory instead of temp directory for for
intermediate files during binaries merge.

LiyouZhou commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/TARGET_MUSCA_A1_NS/device/TOOLCHAIN_GCC_ARM/musca_ns.ld:
```diff
@@ -0,0 +1,212 @@
+        *(.ctors)
+
+        /* .dtors */
+         *crtbegin.o(.dtors)
```
The alignment of indentation of code here is off.

LiyouZhou commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/TARGET_MUSCA_A1_NS/device/TOOLCHAIN_GCC_ARM/musca_ns.ld:
```diff
@@ -0,0 +1,212 @@
+        KEEP(*(SORT(.init_array.*)))
+        KEEP(*(.init_array))
+        PROVIDE_HIDDEN (__init_array_end = .);
+
```
too many empty lines here.

LiyouZhou commented on tools/targets/__init__.py:
```diff
@@ -31,6 +31,7 @@
 from tools.utils import json_file_to_dict, NotSupportedException
 from tools.psa import find_secure_image
 
+
```
don't add random lines.

LiyouZhou commented on targets/TARGET_ARM_SSG/mbed_rtx.h:
```diff
@@ -39,6 +43,22 @@
+#endif  /* TARGET_MUSCA_A1 */
 
 #endif  /* MBED_MBED_RTX_H */
+
```
remove empty new line

LiyouZhou commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/TARGET_MUSCA_A1_S/spm_hal.c:
```diff
@@ -0,0 +1,324 @@
+    NVIC_ClearPendingIRQ(S_MPC_COMBINED_IRQn);
+
+    /* Print fault message and block execution */
+    LOG_MSG("Oops... MPC fault!!!");
```
can we have something less oopsy

LiyouZhou commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/TARGET_MUSCA_A1_S/spm_hal.c:
```diff
@@ -0,0 +1,324 @@
+    NVIC_ClearPendingIRQ(S_PPC_COMBINED_IRQn);
+
+    /* Print fault message*/
+    LOG_MSG("Oops... PPC fault!!!");
```
less oopsy

LiyouZhou commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/gpio_api.c:
```diff
@@ -0,0 +1,81 @@
+
+uint32_t gpio_set(PinName pin)
+{
+   /* Due to a HW limitation, GPIO in Musca-A1 is Secure only, so secure
```
please use 4 spaces indentation in this file.

LiyouZhou commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/TARGET_MUSCA_A1_S/flash_api.c:
```diff
@@ -0,0 +1,132 @@
+
+#define FLASH_DEV          MT25QL_DEV_NS
+
+int32_t flash_init(flash_t *obj)
```
this function can be called multiple times without calling de-init, can the driver handle that? Could we add a init flag? Same questions for de-init as well.

0xc0170 replied:
Good question ,we will need to update flash HAL to add defined/undefined behavior. 

Current driver has a note "Should be called once per lifetime of the object." so this HAL does not do any check, a caller should do.

LiyouZhou commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/TARGET_MUSCA_A1_S/flash_api.c:
```diff
@@ -0,0 +1,132 @@
+uint32_t flash_get_start_address(const flash_t *obj)
+{
+    (void)(obj);
+    return NS_QSPI_ALIAS_BASE;
```
why are we returning NS base if this is the secure side?

urutva replied:
Following definitions are available in flash_layout.h under TARGET_MUSCA_A1/partition/
```#define S_QSPI_ALIAS_BASE               (0x10200000)
#define NS_QSPI_ALIAS_BASE              (0x00200000)
```

LiyouZhou commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/TARGET_MUSCA_A1_NS/prebuilt/README.md:
````diff
@@ -0,0 +1,43 @@
+## tfm.bin
+
+```sh
+mbed compile -t GCC_ARM -m ARM_MUSCA_A1_S --app-config tools/psa/tfm/mbed_app.json --profile release
````
please add version information or the expected mbed-os version, so that the binary is reproducible.

LiyouZhou commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/TARGET_MUSCA_A1_NS/prebuilt/README.md:
````diff
@@ -0,0 +1,43 @@
+mbed compile -t GCC_ARM -m ARM_MUSCA_A1_S --app-config tools/psa/tfm/mbed_app.json --profile release
+```
+
+## spm_smoke.bin
````
do we need all these test binaries? They add 1.5mb to the clone size, given how mbed-cli works this is not desirable at all.

0xc0170 replied:
That is how it is being done so far, we shall revisit this in future (not the PRs scope). 

cc @ARMmbed/mbed-os-tools  FYI

bulislaw reviewed (approved):
@Patater sorry! I miss-understood the situation here, I'm aware about IAR and TF-M status. In this case we should get it in. @ARMmbed/mbed-os-maintainers please note someone should review the code properly, I just skimmed the diff.

0xc0170 committed:
gpio: indent fix

0xc0170 committed:
ARM_SSG: mbed_rtx remove empty line

urutva commented on TESTS/psa/its_ps/main.cpp:
```diff
@@ -178,8 +178,10 @@ utest::v1::status_t case_its_setup_handler(const Case *const source, const size_
 Case cases[] = {
     Case("PSA prot internal storage - Basic", case_its_setup_handler<its>, pits_ps_test<its>, case_its_teardown_handler),
     Case("PSA prot internal storage - Write-once", case_its_setup_handler<its>, pits_ps_write_once_test<its>, case_its_teardown_handler),
+#if COMPONENT_FLASHIAP
```
Shouldn't the `its (internal trusted storage)` tests be wrapped with `#if COMPONENT_FLASHIAP` instead of `ps` (external protected storage)?
@orenc17 suggestion?

orenc17 replied:
in MUSCA A1 we dont have access to the flash for NS
protected storage is set to use flash storage on the non-secure domain

this is a workaround for this board

urutva commented on features/storage/kvstore/conf/global/mbed_lib.json:
```diff
@@ -34,6 +34,9 @@
         "LPC55S69_S": {
             "storage_type": "TDB_INTERNAL"
+        },
+        "MUSCA_A1_S": {
```
Shouldn't this be `ARM_MUSCA_A1_S` as targets.json only contains `ARM_MUSCA_A1_S` as the target and not `MUSCA_A1_S `?

Commit:
CR changes

Commit:
Update secure binaries
